File: python/src/browser_commander/browser/media.py
"""Media emulation for browser-commander.

Provides unified color scheme emulation across Playwright and Selenium.
"""

from __future__ import annotations

from typing import Any, Literal

ColorScheme = Literal["light", "dark", "no-preference"]

VALID_COLOR_SCHEMES = ("light", "dark", "no-preference")


async def emulate_media(
    page: Any,
    engine: str,
    color_scheme: ColorScheme | None = None,
) -> None:
    """Emulate media features (e.g. prefers-color-scheme) for the page.

    Args:
        page: Playwright or Selenium page/driver object.
        engine: Engine type: 'playwright' or 'selenium'.
        color_scheme: Color scheme to emulate: 'light', 'dark', 'no-preference', or None to reset.

    Raises:
        ValueError: If page or engine is not provided, or if color_scheme is invalid.
        ValueError: If the engine is not supported.
    """
    if page is None:
        msg = "page is required in emulate_media"
        raise ValueError(msg)
    if not engine:
        msg = "engine is required in emulate_media"
        raise ValueError(msg)

    if color_scheme is not None and color_scheme not in VALID_COLOR_SCHEMES:
        msg = (
            f'Invalid color_scheme: "{color_scheme}". '
            f"Expected one of: {', '.join(VALID_COLOR_SCHEMES)}, or None"
        )
        raise ValueError(msg)

    if engine == "playwright":
        # Playwright supports emulate_media natively
        media_options: dict[str, Any] = {}
        if color_scheme is not None:
            media_options["color_scheme"] = color_scheme
        await page.emulate_media(**media_options)

    elif engine == "selenium":
        # Selenium uses Chrome DevTools Protocol (CDP) for media emulation
        if color_scheme is None:
            # Reset to default (empty string resets the emulation)
            page.execute_cdp_cmd(
                "Emulation.setEmulatedMedia",
                {
                    "features": [
                        {"name": "prefers-color-scheme", "value": ""},
                    ]
                },
            )
        else:
            page.execute_cdp_cmd(
                "Emulation.setEmulatedMedia",
                {
                    "features": [
                        {"name": "prefers-color-scheme", "value": color_scheme},
                    ]
                },
            )

    else:
        msg = f"Unsupported engine: {engine}. Expected 'playwright' or 'selenium'"
        raise ValueError(msg)

File: js/src/browser/media.js
/**
 * Browser Commander - Media Emulation
 * Provides unified color scheme emulation across Playwright and Puppeteer.
 */

const VALID_COLOR_SCHEMES = ['light', 'dark', 'no-preference'];

/**
 * Emulate media features (e.g. prefers-color-scheme) for the page.
 *
 * @param {Object} options
 * @param {Object} options.page - Playwright or Puppeteer page object
 * @param {string} options.engine - Engine type: 'playwright' or 'puppeteer'
 * @param {string|null} [options.colorScheme] - Color scheme: 'light', 'dark', 'no-preference', or null to reset
 * @returns {Promise<void>}
 */
export async function emulateMedia({ page, engine, colorScheme } = {}) {
  if (!page) {
    throw new Error('page is required in emulateMedia');
  }
  if (!engine) {
    throw new Error('engine is required in emulateMedia');
  }

  if (colorScheme !== null && colorScheme !== undefined) {
    if (!VALID_COLOR_SCHEMES.includes(colorScheme)) {
      throw new Error(
        `Invalid colorScheme: "${colorScheme}". Expected one of: ${VALID_COLOR_SCHEMES.join(', ')}, or null`
      );
    }
  }

  if (engine === 'playwright') {
    // Playwright supports emulateMedia natively
    const mediaOptions = {};
    if (colorScheme !== undefined) {
      mediaOptions.colorScheme = colorScheme;
    }
    await page.emulateMedia(mediaOptions);
  } else if (engine === 'puppeteer') {
    // Puppeteer supports emulateMediaFeatures since v5.4.0
    if (colorScheme === null || colorScheme === undefined) {
      // Reset to default
      await page.emulateMediaFeatures([
        { name: 'prefers-color-scheme', value: '' },
      ]);
    } else {
      await page.emulateMediaFeatures([
        { name: 'prefers-color-scheme', value: colorScheme },
      ]);
    }
  } else {
    throw new Error(
      `Unsupported engine: ${engine}. Expected 'playwright' or 'puppeteer'`
    );
  }
}
